src/tsickle_compiler_host.ts

```diff
@@ -0,0 +1,259 @@
+import * as path from 'path';
+import {SourceMapConsumer, SourceMapGenerator} from 'source-map';
+import * as ts from 'typescript';
+
+import {convertDecorators} from './decorator-annotator';
+import {processES5} from './es5processor';
+import {ModulesManifest} from './modules_manifest';
+import {annotate} from './tsickle';
+
+export enum Pass {
+  None,
+  DecoratorDownlevel,
+  Tsickle
+}
+
+export interface TsickleCompilerHostOptions {
+  googmodule: boolean;
+  es5Mode: boolean;
+  tsickleTyped: boolean;
+  prelude: string;
+}
+
+export interface TsickleEnvironment {
+  /**
+   * If true, tsickle and decorator downlevel processing will be skipped for
+   * that file.
+   */
+  shouldSkipTsickleProcessing: (fileName: string) => boolean;
+  /**
+   * Takes a context (the current file) and the path of the file to import
+   *  and generates a googmodule module name
+   */
+  pathToModuleName: (context: string, importPath: string) => string;
+  /**
+   * Tsickle treats warnings as errors, if true, ignore warnings.  This might be
+   * useful for e.g. third party code.
+   */
+  shouldIgnoreWarningsForPath: (filePath: string) => boolean;
+  /**
+   * If we do googmodule processing, we polyfill module.id, since that's
+   * part of ES6 modules.  This function determines what the module.id will be
+   * for each file.
+   */
+  fileNameToModuleId: (fileName: string) => string;
+}
+
+/**
+ * TsickleCompilerHost does tsickle processing of input files, including
+ * closure type annotation processing, decorator downleveling and
+ * require -> googmodule rewriting.
+ */
+export class TsickleCompilerHost implements ts.CompilerHost {
+  private ANNOTATION_SUPPORT = `
+interface DecoratorInvocation {
+  type: Function;
+  args?: any[];
+}
+`;
+  // The manifest of JS modules output by the compiler.
+  public modulesManifest: ModulesManifest = new ModulesManifest();
+
+  /** Error messages produced by tsickle, if any. */
+  public diagnostics: ts.Diagnostic[] = [];
+
+  /** externs.js files produced by tsickle, if any. */
+  public externs: {[fileName: string]: string} = {};
+
+  private decoratorDownlevelSourceMaps = new Map<string, SourceMapGenerator>();
+  private tsickleSourceMaps = new Map<string, SourceMapGenerator>();
+
+  constructor(
+      private delegate: ts.CompilerHost, private options: TsickleCompilerHostOptions,
+      private environment: TsickleEnvironment, private oldProgram?: ts.Program,
+      private pass = Pass.None) {}
+
+
+  public reconfigureForRun(program: ts.Program, pass: Pass) {
+    this.oldProgram = program;
+    this.pass = pass;
+  }
+
+  getSourceFile(
+      fileName: string, languageVersion: ts.ScriptTarget,
+      onError?: (message: string) => void): ts.SourceFile {
+    if (this.pass === Pass.None) {
+      return this.delegate.getSourceFile(fileName, languageVersion, onError);
+    }
+
+    if (!this.oldProgram) {
+      throw new Error('tried to run a pass other than None without setting a program');
```

Comment:
The real problem here is that I want the constructor to accept either a pass and an oldProgram or neither.  I've found a way to ensure that happens, so I've removed the check and error.